# MAG catalogue

```{r load_data_mag, message=FALSE, warning=FALSE}
load("data/data.Rdata")
```

## filter samples with high host data
```{r load_data_mag_filt, comment="", message=FALSE, warning=FALSE, eval=FALSE}
sample_metadata <- sample_metadata%>%
  filter(!sample %in% c("EHI02721", "EHI02712", "EHI02700", "EHI02720", "EHI02749", "EHI02719", "EHI02729", "EHI02715", "EHI02722"))

genome_counts_filt <- genome_counts %>% 
    select(one_of(c("genome",sample_metadata$sample)))%>%
  filter(rowSums(. != 0, na.rm = TRUE) > 0) %>%
  select_if(~!all(. == 0))

genome_metadata <- genome_metadata %>% 
  semi_join(., genome_counts_filt, by = "genome") %>% 
  arrange(match(genome,genome_counts_filt$genome))

genome_tree <- keep.tip(genome_tree, tip=genome_metadata$genome) # keep only MAG tips

#load("data/genome_gifts.Rdata")
```

## Genome phylogeny

```{r genome_phylogeny, message=FALSE, warning=FALSE, fig.height=10, fig.width=10, fig.fullwidth=TRUE}
# Generate the phylum color heatmap
phylum_heatmap <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
    right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
    select(genome,phylum) %>%
    mutate(phylum = factor(phylum, levels = unique(phylum))) %>%
    column_to_rownames(var = "genome")

newspecies_table <- genome_metadata %>% 
  mutate(newspecies=ifelse(species=="s__","Y","N")) %>% 
  select(genome,newspecies) %>% 
  column_to_rownames(var = "genome")
  

# Generate  basal tree
circular_tree <- force.ultrametric(genome_tree, method="extend") %>% # extend to ultrametric for the sake of visualisation
    ggtree(., layout="fan", open.angle=10, size=0.2)

# Add phylum ring
circular_tree <- gheatmap(circular_tree, phylum_heatmap, offset=0.1, width=0.1, colnames=FALSE) +
        scale_fill_manual(values=phylum_colors) +
        theme(legend.position = "none", plot.margin = margin(0, 0, 0, 0), panel.margin = margin(0, 0, 0, 0))

# Flush color scale to enable a new color scheme in the next ring
circular_tree <- circular_tree + new_scale_fill()

# Add completeness ring
circular_tree <- circular_tree +
        new_scale_fill() +
        scale_fill_gradient(low = "#d1f4ba", high = "#f4baba") +
        geom_fruit(
                data=genome_metadata,
                geom=geom_bar,
                mapping = aes(x=completeness, y=genome, fill=contamination),
                offset = 0.30,
                orientation="y",
              stat="identity")

# Add genome-size ring
circular_tree <- circular_tree + new_scale_fill()

circular_tree <- gheatmap(circular_tree, newspecies_table, offset=1.5, width=0.1, colnames=FALSE) +
        scale_fill_manual(values=c("#f4f4f4","#666666"))
        theme(legend.position = "none", plot.margin = margin(0, 0, 0, 0), panel.margin = margin(0, 0, 0, 0))

# Add text
circular_tree <-  circular_tree +
        annotate('text', x=3.1, y=0, label='            Phylum', family='arial', size=3.5) +
        annotate('text', x=4, y=0, label='                         Genome quality', family='arial', size=3.5) +
        annotate('text', x=4.6, y=0, label='                     New species', family='arial', size=3.5)

#Plot circular tree
circular_tree %>% open_tree(30) %>% rotate_tree(90)
```

## Taxonomy overview

```{r genome_taxonomy_bacteria, message=FALSE, warning=FALSE,}
tax_mag <-genome_metadata %>% 
  group_by(phylum) %>%
  summarise(mag_n=n()) 

tax_mag %>%
  mutate(percetage_mag=round(mag_n*100/sum(mag_n), 2)) %>% 
  arrange(-percetage_mag) %>%
  tt()
```

## Mag size (MB)

```{r}
genome_metadata <-genome_metadata %>% 
  mutate(corrected_size=100*length/completeness)
```

Mags average size (MB)
```{r mag_size_all_mean, message=FALSE, warning=FALSE}
genome_metadata %>% 
  summarise(Average_corrected_size=mean(corrected_size))
```

Minimum Mags size (MB)

```{r mag_size_min,message=FALSE, warning=FALSE}
genome_metadata %>% 
  filter(corrected_size==min(corrected_size))
```

Maximum Mags size (MB)

```{r mag_size_max,message=FALSE, warning=FALSE}
genome_metadata %>% 
  filter(corrected_size==max(corrected_size))
```

Mags arrange by size (MB)
```{r mag_size_all, message=FALSE, warning=FALSE}
genome_metadata %>% 
  arrange(corrected_size) %>% 
  tt()
```

Mags average size and completeness by phylum

```{r mag_size_phylum, message=FALSE, warning=FALSE}
genome_metadata %>% 
  group_by(phylum) %>%
  summarise(average_size=mean(corrected_size),sd_size=sd(corrected_size),average_comp=mean(completeness),sd_comp=sd(completeness)) %>%
  mutate(Size=str_c(round(average_size,2),"±",round(sd_size,2)),
         Completeness=str_c(round(average_comp,2),"±",round(sd_comp,2))) %>% 
  arrange(-average_size) %>% 
  select(phylum, Size, Completeness)
```

Mags average size and completeness by phylum in wild

```{r mag_size_phylum_wild, message=FALSE, warning=FALSE}

wild_data <-sample_metadata %>% 
  filter(diet=="Wild")
  
  
wild_genome <- genome_counts %>% 
  select(genome, all_of(wild_data$sample)) %>%
  filter(rowSums(across(where(is.numeric)))!=0) 
  

wild_genome_metadata <- genome_metadata %>% 
  filter(genome %in% wild_genome$genome) %>% 
  arrange(match(genome,wild_genome$genome))

wild_genome_metadata %>% 
  group_by(phylum) %>%
  summarise(average_size=mean(corrected_size),sd_size=sd(corrected_size),average_comp=mean(completeness),sd_comp=sd(completeness)) %>%
  mutate(Size=str_c(round(average_size,2),"±",round(sd_size,2)),
         Completeness=str_c(round(average_comp,2),"±",round(sd_comp,2))) %>% 
  arrange(-average_size) %>% 
  select(phylum, Size, Completeness)

```

```{r mag_size_wild_mean, message=FALSE, warning=FALSE}
wild_genome_metadata %>% 
  summarise(average_size=mean(corrected_size),sd_size=sd(corrected_size),average_comp=mean(completeness),sd_comp=sd(completeness)) %>%
  mutate(Size=str_c(round(average_size,2),"±",round(sd_size,2)),
         Completeness=str_c(round(average_comp,2),"±",round(sd_comp,2))) %>% 
  arrange(-average_size) %>% 
  select(-Size, -Completeness)
```

Mags average size and completeness by phylum in captivity

```{r mag_size_phylum_capt, message=FALSE, warning=FALSE}

pre_data <-sample_metadata %>% 
  filter(diet=="Pre_grass")
  
  
pre_genome <- genome_counts %>% 
  select(genome, all_of(pre_data$sample)) %>%
  filter(rowSums(across(where(is.numeric)))!=0) 
  

pre_genome_metadata <- genome_metadata %>% 
  filter(genome %in% pre_genome$genome) %>% 
  arrange(match(genome,pre_genome$genome))

pre_genome_metadata %>% 
  group_by(phylum) %>%
  summarise(average_size=mean(corrected_size),sd_size=sd(corrected_size),average_comp=mean(completeness),sd_comp=sd(completeness)) %>%
  mutate(Size=str_c(round(average_size,2),"±",round(sd_size,2)),
         Completeness=str_c(round(average_comp,2),"±",round(sd_comp,2))) %>% 
  arrange(-average_size) %>% 
  select(phylum, Size, Completeness)

```

```{r mag_size_mean_cap, message=FALSE, warning=FALSE}
pre_genome_metadata %>% 
  summarise(average_size=mean(corrected_size),sd_size=sd(corrected_size),average_comp=mean(completeness),sd_comp=sd(completeness)) %>%
  mutate(Size=str_c(round(average_size,2),"±",round(sd_size,2)),
         Completeness=str_c(round(average_comp,2),"±",round(sd_comp,2))) %>% 
  arrange(-average_size) %>% 
  select(-Size, -Completeness)
```

## Genome quality

```{r genome_quality, message=FALSE, warning=FALSE}
genome_metadata %>% 
    summarise(completeness_mean=mean(completeness) %>% round(2) %>% as.character(), 
              completeness_sd=sd(completeness) %>% round(2) %>% as.character(), 
              contamination_mean=mean(contamination) %>% round(2), 
              contamination_sd=sd(contamination) %>% round(2)) %>%
    unite("Completeness",completeness_mean, completeness_sd, sep = " ± ", remove = TRUE) %>%
    unite("Contamination",contamination_mean, contamination_sd, sep = " ± ", remove = TRUE) %>%
    tt()
```

```{r genome_quality_plot, message=FALSE, warning=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}

#Generate quality biplot
genome_biplot <- genome_metadata %>%
  select(c(genome,domain,phylum,completeness,contamination,length)) %>%
  arrange(match(genome, rev(genome_tree$tip.label))) %>% #sort MAGs according to phylogenetic tree
  ggplot(aes(x=completeness,y=contamination,size=length,color=phylum)) +
              geom_point(alpha=0.7) +
                    ylim(c(10,0)) +
                    scale_color_manual(values=phylum_colors) +
                    labs(y= "Contamination", x = "Completeness") +
                    theme_classic() +
                    theme(legend.position = "none")

#Generate contamination boxplot
genome_contamination <- genome_metadata %>%
            ggplot(aes(y=contamination)) +
                    ylim(c(10,0)) +
                    geom_boxplot(colour = "#999999", fill="#cccccc") +
                    theme_void() +
                    theme(legend.position = "none",
                        axis.title.x = element_blank(),
                        axis.title.y = element_blank(),
                        axis.text.y=element_blank(),
                        axis.ticks.y=element_blank(),
                        axis.text.x=element_blank(),
                        axis.ticks.x=element_blank(),
                        plot.margin = unit(c(0, 0, 0.40, 0),"inches")) #add bottom-margin (top, right, bottom, left)

#Generate completeness boxplot
genome_completeness <- genome_metadata %>%
        ggplot(aes(x=completeness)) +
                xlim(c(50,100)) +
                geom_boxplot(colour = "#999999", fill="#cccccc") +
                theme_void() +
                theme(legend.position = "none",
                    axis.title.x = element_blank(),
                    axis.title.y = element_blank(),
                    axis.text.y=element_blank(),
                    axis.ticks.y=element_blank(),
                    axis.text.x=element_blank(),
                    axis.ticks.x=element_blank(),
                    plot.margin = unit(c(0, 0, 0, 0.50),"inches")) #add left-margin (top, right, bottom, left)

#Render composite figure
grid.arrange(grobs = list(genome_completeness,genome_biplot,genome_contamination),
        layout_matrix = rbind(c(1,1,1,1,1,1,1,1,1,1,1,4),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3)))

```


## Functional overview

```{r annotation_stats, message=FALSE, warning=FALSE, fig.height=10, fig.width=10, fig.fullwidth=TRUE}
genome_annotations <- read_tsv("data/genome_annotations.tsv.xz") %>%
    rename(gene=1, genome=2, contig=3)

# Predicted genes
genome_annotations %>%
  nrow()

# Some annotation
genome_annotations %>%
  filter(if_any(c(kegg_id, pfam_hits, cazy_hits), ~ !is.na(.))) %>% 
  nrow()

# KEGG annotation
genome_annotations %>%
  filter(!is.na(kegg_id)) %>% 
  nrow()
```

```{r function_heatmap, message=FALSE, warning=FALSE, fig.height=10, fig.width=10, fig.fullwidth=TRUE}

# Aggregate basal GIFT into elements
genome_gifts<-genome_gifts_raw
function_table <- genome_gifts %>%
    to.elements(., GIFT_db)

# Generate  basal tree
function_tree <- force.ultrametric(genome_tree, method="extend") %>%
                ggtree(., size = 0.3) 

#Add phylum colors next to the tree tips
function_tree <- gheatmap(function_tree, phylum_heatmap, offset=0, width=0.1, colnames=FALSE) +
            scale_fill_manual(values=phylum_colors) +
            labs(fill="Phylum")

#Reset fill scale to use a different colour profile in the heatmap
function_tree <- function_tree + new_scale_fill()

#Add functions heatmap
function_tree <- gheatmap(function_tree, function_table, offset=0.5, width=3.5, colnames=FALSE) +
            vexpand(.08) +
            coord_cartesian(clip = "off") +
            scale_fill_gradient(low = "#f4f4f4", high = "steelblue", na.value="white") +
            labs(fill="GIFT")

#Reset fill scale to use a different colour profile in the heatmap
function_tree <- function_tree + new_scale_fill()

# Add completeness barplots
function_tree <- function_tree +
            geom_fruit(data=genome_metadata,
            geom=geom_bar,
            grid.params=list(axis="x", text.size=2, nbreak = 1),
            axis.params=list(vline=TRUE),
            mapping = aes(x=length, y=genome, fill=completeness),
                 offset = 3.8,
                 orientation="y",
                 stat="identity") +
            scale_fill_gradient(low = "#cf8888", high = "#a2cc87") +
            labs(fill="Genome\ncompleteness")

function_tree
```

## Functional ordination

```{r function_ordination, message=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
# Generate the tSNE ordination
tSNE_function <- Rtsne(X=function_table, dims = 2, check_duplicates = FALSE)

# Plot the ordination
function_ordination <- tSNE_function$Y %>%
                as.data.frame() %>%
                mutate(genome=rownames(function_table)) %>%
                inner_join(genome_metadata, by="genome") %>%
                rename(tSNE1="V1", tSNE2="V2") %>%
                select(genome,phylum,tSNE1,tSNE2, length) %>%
                ggplot(aes(x = tSNE1, y = tSNE2, color = phylum, size=length))+
                            geom_point(shape=16, alpha=0.7) +
                            scale_color_manual(values=phylum_colors) +
                            theme_minimal() +
                labs(color="Phylum", size="Genome size") +
                guides(color = guide_legend(override.aes = list(size = 5))) # enlarge Phylum dots in legend

function_ordination
```