
# Functional differences

```{r load_data_func,comment="", echo=FALSE, message=FALSE, warning=FALSE}
load("data/data_host_filtered.Rdata")
load("data/ancombc_filtered_functional.Rdata")
```

```{r gift, comment="", echo=FALSE, message=FALSE, warning=FALSE}
genome_counts_filt <- genome_counts_filt[genome_counts_filt$genome %in% rownames(genome_gifts_raw),]
genome_counts_filt <- genome_counts_filt %>%
#  remove_rownames() %>% 
  column_to_rownames(., "genome") %>%
  filter(rowSums(. != 0, na.rm = TRUE) > 0) %>%
  select_if(~!all(. == 0))%>%
  rownames_to_column(., "genome")

genome_gifts <- genome_gifts_raw[rownames(genome_gifts_raw) %in% genome_counts_filt$genome,]
genome_gifts <- genome_gifts[, colSums(genome_gifts != 0) > 0]

# genome_counts_filt <- genome_counts_filt[genome_counts_filt$genome %in% rownames(genome_gifts),] 
# rownames(genome_counts_filt) <- NULL
#Aggregate bundle-level GIFTs into the compound level
GIFTs_elements <- to.elements(genome_gifts,GIFT_db)
GIFTs_elements_filtered <- GIFTs_elements[rownames(GIFTs_elements) %in% genome_counts_filt$genome,]
GIFTs_elements_filtered <- as.data.frame(GIFTs_elements_filtered) %>% 
  select_if(~ !is.numeric(.) || sum(.) != 0)

#Aggregate element-level GIFTs into the function level
GIFTs_functions <- to.functions(GIFTs_elements_filtered,GIFT_db)

#Aggregate function-level GIFTs into overall Biosynthesis, Degradation and Structural GIFTs
GIFTs_domains <- to.domains(GIFTs_functions,GIFT_db)

#Get community-weighed average GIFTs per sample
genome_counts_row <- genome_counts_filt %>%
  mutate_at(vars(-genome),~./sum(.)) %>% 
  column_to_rownames(., "genome") 
# genome_counts_row <- rownames_to_column(genome_counts_row, "genome")
GIFTs_elements_community <- to.community(GIFTs_elements_filtered,genome_counts_row,GIFT_db)
GIFTs_functions_community <- to.community(GIFTs_functions,genome_counts_row,GIFT_db)
GIFTs_domains_community <- to.community(GIFTs_domains,genome_counts_row,GIFT_db)
```

```{r gift, comment="", echo=FALSE, message=FALSE, warning=FALSE, eval=FALSE}
gift_meta <- GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>%
  left_join(genome_metadata[,1:8], by = join_by(genome == genome)) %>% 
  group_by(phylum)  %>%
  summarize(across(where(is.numeric), mean, na.rm = TRUE))%>%
    pivot_longer(!phylum,names_to="trait",values_to="gift") %>%
    mutate(functionid = substr(trait, 1, 3)) %>%
    mutate(trait = case_when(
      trait %in% GIFT_db$Code_element ~ GIFT_db$Element[match(trait, GIFT_db$Code_element)],
      TRUE ~ trait
    )) %>%
    mutate(functionid = case_when(
      functionid %in% GIFT_db$Code_function ~ GIFT_db$Function[match(functionid, GIFT_db$Code_function)],
      TRUE ~ functionid
    )) %>%
    mutate(trait=factor(trait,levels=unique(GIFT_db$Element))) %>%
    mutate(functionid=factor(functionid,levels=unique(GIFT_db$Function))) 

gift_meta %>% 
  group_by(phylum, functionid) %>% 
  summarise(mean=mean(gift)) %>% 
  filter(functionid=="Xenobiotic degradation") %>% 
  arrange(-mean)

gift_meta %>% 
  group_by(phylum, trait) %>% 
  summarise(mean=mean(gift)) %>% 
  filter(trait=="D-Xylose") %>% 
  arrange(-mean)

genomes_funciton <- GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column("Code_element")  %>% 
  left_join(GIFT_db, by = join_by(Code_element == Code_element)) %>% 
  filter(Element=="Oxacillin") %>% 
  select_if(~ . > 0) %>% 
  select(2)%>%
  colnames()
genome_metadata %>% 
  filter(genome %in% genomes_funciton) %>% 
  arrange(phylum)

genomes_funciton <- GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column("Code_element")  %>% 
  left_join(GIFT_db, by = join_by(Code_element == Code_element)) %>% 
  filter(Element=="Cephalosporin") %>% 
  select_if(~ . > 0) %>% 
  select(2:4)%>%
  colnames()
genome_metadata %>% 
  filter(genome %in% genomes_funciton) %>% 
  arrange(phylum)

genomes_funciton <- GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column("Code_element")  %>% 
  left_join(GIFT_db, by = join_by(Code_element == Code_element)) %>% 
  filter(Element=="Mercury") %>% 
  select_if(~ . > 0) %>% 
  select(2)%>%
  colnames()
genome_metadata %>% 
  filter(genome %in% genomes_funciton) %>% 
  arrange(phylum)
#Cumate
genomes_funciton <- GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column("Code_element")  %>% 
  left_join(GIFT_db, by = join_by(Code_element == Code_element)) %>% 
  filter(Element=="Cumate") %>% 
  select_if(~ . > 0) %>% 
  select(2)%>%
  colnames()
genome_metadata %>% 
  filter(genome %in% genomes_funciton) %>% 
  arrange(phylum)
#Anthranilate
genomes_funciton <- GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column("Code_element")  %>% 
  left_join(GIFT_db, by = join_by(Code_element == Code_element)) %>% 
  filter(Element=="Anthranilate") %>% 
  select_if(~ . > 0) %>% 
  select(2)%>%
  colnames()
genome_metadata %>% 
  filter(genome %in% genomes_funciton) %>% 
  arrange(phylum)
#aerobactin
genomes_funciton <- GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column("Code_element")  %>% 
  left_join(GIFT_db, by = join_by(Code_element == Code_element)) %>% 
  filter(Element=="Aerobactin") %>% 
  select(where(~ sum(. > 0) > 0))  %>%
  select(2)%>%
  colnames()
genome_metadata %>% 
  filter(genome %in% genomes_funciton) %>% 
  arrange(phylum)
#betaine
genomes_function <- GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column("Code_element")  %>% 
  left_join(GIFT_db, by = join_by(Code_element == Code_element)) %>%
    filter(Function=="Nitrogen compound degradation") %>%
  filter(Element=="Betaine") %>% 
  select_if(~ . > 0) %>% 
  select(2:8)%>%
  colnames()
genome_metadata %>% 
  filter(genome %in% genomes_function) %>% 
  arrange(phylum)
#Valine, Arginine
genomes_function <- GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column("Code_element")  %>% 
  left_join(GIFT_db, by = join_by(Code_element == Code_element)) %>%
    filter(Function=="Amino acid degradation") %>%
  filter(Element=="Arginine") %>% 
  select(where(~ sum(. > 0) > 0))
#Captive
#Streptothricin, Carbapenem, Lincosamide and Streptogramin
genomes_function <- GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column("Code_element")  %>% 
  left_join(GIFT_db, by = join_by(Code_element == Code_element)) %>%
    filter(Function=="Antibiotic degradation") %>%
  filter(Element=="Streptothricin")%>% 
  select_if(~ . > 0) %>% 
  select(2:3)%>%
  colnames()
genome_metadata %>% 
  filter(genome %in% genomes_function) %>% 
  arrange(phylum)
genomes_function <- GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column("Code_element")  %>% 
  left_join(GIFT_db, by = join_by(Code_element == Code_element)) %>%
    filter(Function=="Antibiotic degradation") %>%
  filter(Element=="Carbapenem")%>% 
  select_if(~ . > 0) %>% 
  select(2:7)%>%
  colnames()
genome_metadata %>% 
  filter(genome %in% genomes_function) %>% 
  arrange(phylum)
#Lincosamide
genomes_function <- GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column("Code_element")  %>% 
  left_join(GIFT_db, by = join_by(Code_element == Code_element)) %>%
    filter(Function=="Antibiotic degradation") %>%
  filter(Element=="Lincosamide")%>% 
  select_if(~ . > 0) %>% 
  select(2:94)%>%
  colnames()
genome_metadata %>% 
  filter(genome %in% genomes_function) %>% 
  arrange(phylum)

genome_counts_filt %>%
  mutate_at(vars(-genome),~./sum(.))%>%
  filter(genome %in% genomes_function) %>% 
  pivot_longer(!genome, names_to = "sample", values_to = "abundance") %>% 
  left_join(sample_metadata, by=join_by(sample==sample)) %>% 
  filter(diet!="Post_grass") %>% 
  ggplot(aes(y = abundance, x = diet, group=diet, color=diet, fill=diet)) +
  geom_jitter(width = 0.2, show.legend = FALSE) +
  geom_boxplot(width = 0.5, alpha=0.5,outlier.shape = NA, show.legend = FALSE) +
  scale_color_manual(values=diet_colors)+
  scale_fill_manual(values=diet_colors) +
  scale_x_discrete(labels=c("Wild" = "Wild", "Pre_grass" = "Captive-bred")) +
  coord_cartesian(xlim = c(1, NA)) +
  theme_classic() +
  theme(
    strip.background = element_blank(),
    panel.grid.minor.x = element_line(size = .1, color = "grey"),
    axis.text = element_text(size=10),
    axis.title.y = element_text(margin = margin(t = 0, r = 20, b = 0, l = 0)),
    axis.title.x = element_text(margin = margin(t = 20, r = 0, b = 0, l = 0))
    )

genomes_function <- GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column("Code_element")  %>% 
  left_join(GIFT_db, by = join_by(Code_element == Code_element)) %>%
    filter(Function=="Sugar degradation") %>%
  filter(Element=="D-Xylose")%>% 
  select(where(~ max(.) == 1)) %>% 
  select(1:37)%>%
  colnames()
genome_metadata %>% 
  filter(genome %in% genomes_function) %>% 
  arrange(phylum)


a <- GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column("Code_element")  %>% 
  left_join(GIFT_db, by = join_by(Code_element == Code_element)) %>% 
  filter(Function=="Sugar degradation") %>%
  filter(Element=="D-Xylose") %>%
  select(-Code_element) %>% 
  select(1:765) %>%
# select(where(~ sum(. > 1) > 0)) %>% 
  select(where(~ max(.) == 1))  %>%
  colnames()

genome_metadata %>% 
  filter(genome %in% a) %>% 
  arrange(phylum)
```
```{r}
GIFT_db %>% 
  filter(Element=="Streptogramin")
  
```


```{r phylum_elem_capacity, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=12, fig.width=10, fig.fullwidth=TRUE}
GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>%
  left_join(genome_metadata[,1:8], by = join_by(genome == genome)) %>% group_by(phylum)  %>%
  summarize(across(where(is.numeric), mean, na.rm = TRUE))%>%
    pivot_longer(!phylum,names_to="trait",values_to="gift") %>%
    mutate(functionid = substr(trait, 1, 3)) %>%
    mutate(trait = case_when(
      trait %in% GIFT_db$Code_element ~ GIFT_db$Element[match(trait, GIFT_db$Code_element)],
      TRUE ~ trait
    )) %>%
    mutate(functionid = case_when(
      functionid %in% GIFT_db$Code_function ~ GIFT_db$Function[match(functionid, GIFT_db$Code_function)],
      TRUE ~ functionid
    )) %>%
    mutate(trait=factor(trait,levels=unique(GIFT_db$Element))) %>%
    mutate(functionid=factor(functionid,levels=unique(GIFT_db$Function))) %>%
    ggplot(aes(x=phylum,y=trait,fill=gift)) +
        geom_tile(colour="white", linewidth=0.2)+
        scale_fill_gradientn(colours=rev(c("#d53e4f", "#f46d43", "#fdae61", "#fee08b", "#e6f598", "#abdda4", "#ddf1da")))+
        facet_grid(functionid ~ ., scales="free",space="free") +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
              axis.text.y = element_text(size=6),
              strip.text.y = element_text(angle = 0)
              ) +
        labs(y="Traits",x="Samples",fill="GIFT")

```

```{r structural_functional_capacity, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=12, fig.width=10, fig.fullwidth=TRUE}
GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>%
  filter(genome %in% struct_mag)  %>% 
    pivot_longer(!genome,names_to="trait",values_to="gift") %>% 
    left_join(struct_mag_meta, by = join_by(genome == genome))%>%
    mutate(functionid = substr(trait, 1, 3)) %>%
    mutate(trait = case_when(
      trait %in% GIFT_db$Code_element ~ GIFT_db$Element[match(trait, GIFT_db$Code_element)],
      TRUE ~ trait
    )) %>%
    mutate(functionid = case_when(
      functionid %in% GIFT_db$Code_function ~ GIFT_db$Function[match(functionid, GIFT_db$Code_function)],
      TRUE ~ functionid
    )) %>%
    mutate(trait=factor(trait,levels=unique(GIFT_db$Element))) %>%
    mutate(functionid=factor(functionid,levels=unique(GIFT_db$Function))) %>%
    ggplot(aes(x=genome,y=trait,fill=gift)) +
        geom_tile(colour="white", linewidth=0.2)+
        scale_fill_gradientn(colours=rev(c("#d53e4f", "#f46d43", "#fdae61", "#fee08b", "#e6f598", "#abdda4", "#ddf1da")))+
        facet_grid(functionid ~ present, scales="free",space="free") +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
              axis.text.y = element_text(size=6),
              strip.text.y = element_text(angle = 0)
              ) +
        labs(y="Traits",x="Samples",fill="GIFT")

```

```{r mci_mean, echo=FALSE, message=FALSE, warning=FALSE}
GIFTs_elements_filtered  %>% 
  rownames_to_column("genome")%>% 
  left_join(genome_metadata[,1:8], by = join_by(genome == genome)) %>% group_by(phylum)  %>%  # Group by phylum
  summarize(across(where(is.numeric), mean, na.rm = TRUE)) %>%  # Calculate the mean per group
  mutate(
    row_mean = rowMeans(across(where(is.numeric)), na.rm = TRUE),  # Calculate row means
    row_sd = apply(across(where(is.numeric)), 1, sd, na.rm = TRUE),
    row_median = apply(across(where(is.numeric)), 1, median, na.rm = TRUE),
    row_max = apply(across(where(is.numeric)), 1, max, na.rm = TRUE),
    row_min = apply(across(where(is.numeric)), 1, min, na.rm = TRUE)# Calculate row standard deviations
  ) %>% 
  select(phylum, row_mean, row_sd, row_median, row_max, row_min) %>% 
  arrange(-row_mean)
```
```{r}
sugar<-GIFT_db %>% 
  filter(Code_function=="D03") %>% 
  select(Code_element) %>% 
  pull()
GIFTs_elements_filtered %>% 
  select(any_of(sugar)) %>%
  rowMeans() %>% 
  as.data.frame() %>% 
  rename(sugar_mean=".") %>% 
  arrange(-sugar_mean) %>%
  rownames_to_column("genome") %>%
  left_join(genome_metadata[,1:8], by = join_by(genome == genome))

GIFTs_elements_filtered %>% 
  select("D0306") %>%
  rowMeans() %>% 
  as.data.frame() %>% 
  rename(sugar_mean=".") %>% 
  arrange(-sugar_mean) %>%
  rownames_to_column("genome") %>%
  left_join(genome_metadata[,1:8], by = join_by(genome == genome))

GIFTs_elements_filtered %>% 
  select("D0307") %>%
  rowMeans() %>% 
  as.data.frame() %>% 
  rename(sugar_mean=".") %>% 
  arrange(-sugar_mean) %>%
  rownames_to_column("genome") %>%
  left_join(genome_metadata[,1:8], by = join_by(genome == genome))

GIFTs_elements_filtered %>% 
  select("D0308") %>%
  rowMeans() %>% 
  as.data.frame() %>% 
  rename(sugar_mean=".") %>% 
  arrange(-sugar_mean) %>%
  rownames_to_column("genome") %>%
  left_join(genome_metadata[,1:8], by = join_by(genome == genome))
  
```

```{r phylum_elem_values, comment="", echo=FALSE, message=FALSE, warning=FALSE}

GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>%
  left_join(genome_metadata[,1:8], by = join_by(genome == genome)) %>% group_by(phylum)  %>%
  summarize(across(where(is.numeric), mean, na.rm = TRUE))%>%
    pivot_longer(!phylum,names_to="trait",values_to="gift") %>%
    mutate(functionid = substr(trait, 1, 3)) %>%
    mutate(trait = case_when(
      trait %in% GIFT_db$Code_element ~ GIFT_db$Element[match(trait, GIFT_db$Code_element)],
      TRUE ~ trait
    )) %>%
    mutate(functionid = case_when(
      functionid %in% GIFT_db$Code_function ~ GIFT_db$Function[match(functionid, GIFT_db$Code_function)],
      TRUE ~ functionid
    )) %>%
    mutate(trait=factor(trait,levels=unique(GIFT_db$Element))) %>%
    mutate(functionid=factor(functionid,levels=unique(GIFT_db$Function))) %>% 
  group_by(phylum, functionid) %>% 
  summarize(median=median(gift))%>%
  pivot_wider(names_from = functionid, values_from = median)
  
```
```{r funct_boxplot, echo=FALSE, message=FALSE, warning=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>%
  left_join(genome_metadata[,1:8], by = join_by(genome == genome)) %>% group_by(phylum)  %>%
  summarize(across(where(is.numeric), mean, na.rm = TRUE))%>%
    pivot_longer(!phylum,names_to="trait",values_to="gift") %>%
    mutate(functionid = substr(trait, 1, 3)) %>%
    mutate(trait = case_when(
      trait %in% GIFT_db$Code_element ~ GIFT_db$Element[match(trait, GIFT_db$Code_element)],
      TRUE ~ trait
    )) %>%
    mutate(functionid = case_when(
      functionid %in% GIFT_db$Code_function ~ GIFT_db$Function[match(functionid, GIFT_db$Code_function)],
      TRUE ~ functionid
    )) %>%
    mutate(trait=factor(trait,levels=unique(GIFT_db$Element))) %>%
    mutate(functionid=factor(functionid,levels=unique(GIFT_db$Function))) %>% 
  group_by(phylum, functionid) %>% 
  summarize(mean=mean(gift))%>%
  ggplot(aes(y = mean, x = functionid, group=functionid, color=phylum)) +
  geom_jitter(width = 0.2, size= 3, show.legend = FALSE) +
  geom_boxplot(width = 0.5, alpha=0.5,outlier.shape = NA, show.legend = FALSE) +
  scale_color_manual(values=phylum_colors)+
#  scale_fill_manual(values=diet_colors) +
#  scale_x_discrete(labels=c("Wild" = "Wild", "Pre_grass" = "Captive-born")) +
#  stat_compare_means(method = "wilcox.test",show.legend = F, size = 3, label.y = c(150), label.x = c(1.5))+
  coord_cartesian(xlim = c(1, NA)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
    strip.background = element_blank(),
    panel.grid.minor.x = element_line(size = .1, color = "grey"),
    axis.text = element_text(size=10),
    axis.title.y = element_text(margin = margin(t = 0, r = 20, b = 0, l = 0)),
    axis.title.x = element_text(margin = margin(t = 20, r = 0, b = 0, l = 0))
    )+
  labs(x = "Origin", y = "Neutral")
```

# Spirochaetota
```{r spiro_func, comment="", echo=FALSE, message=FALSE, warning=FALSE}
spiro_funct <- GIFTs_elements_filtered %>% 
  as.data.frame() %>% 
  rownames_to_column("genome") %>% 
  filter(genome %in% c("EHA04587_bin.1", "EHA04572_bin.6"))%>% 
  select(where(~ max(.) == 1))  %>%
  colnames()

GIFT_db %>% 
  filter(Code_element %in% spiro_funct) %>% 
  tt()


  

GIFTs_elements_filtered %>% 
  select(D0308) %>% 
  arrange(-D0308) %>% 
  rownames_to_column("genome") %>% 
  left_join(genome_metadata, by=join_by(genome==genome))
```

## MCI
```{r gitfs_functional_wild, echo=TRUE,results=TRUE}
GIFTs_functions_community %>%
  rowMeans() %>%
  as_tibble(., rownames = "sample") %>%
  left_join(sample_metadata, by = join_by(sample == sample)) %>%
#  filter(time_point=="0_Wild") %>%
  group_by(diet) %>%
  summarise(MCI = mean(value), sd = sd(value))

MCI <- GIFTs_functions_community %>%
  rowMeans() %>%
  as_tibble(., rownames = "sample") %>%
  left_join(sample_metadata, by = join_by(sample == sample)) 
# %>% 
#   filter(diet!="Post_grass")

shapiro.test(MCI$value)
res.aov <- aov(value ~ diet, data=MCI)
summary(res.aov)
```

## Community level plots
```{r comunity_elem_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
GIFTs_elements_community %>%
    as.data.frame() %>%
    rownames_to_column(var="sample")%>%
    pivot_longer(!sample,names_to="trait",values_to="gift") %>%
    left_join(sample_metadata, by = join_by(sample == sample)) %>%
#  filter(!diet =="Post_grass")%>%
    mutate(functionid = substr(trait, 1, 3)) %>%
    mutate(trait = case_when(
      trait %in% GIFT_db$Code_element ~ GIFT_db$Element[match(trait, GIFT_db$Code_element)],
      TRUE ~ trait
    )) %>%
    mutate(functionid = case_when(
      functionid %in% GIFT_db$Code_function ~ GIFT_db$Function[match(functionid, GIFT_db$Code_function)],
      TRUE ~ functionid
    )) %>%
    mutate(trait=factor(trait,levels=unique(GIFT_db$Element))) %>%
    mutate(functionid=factor(functionid,levels=unique(GIFT_db$Function))) %>%
    ggplot(aes(x=sample,y=trait,fill=gift)) +
        geom_tile(colour="white", linewidth=0.2)+
        scale_fill_gradientn(colours=rev(c("#d53e4f", "#f46d43", "#fdae61", "#fee08b", "#e6f598", "#abdda4", "#ddf1da")))+
        facet_grid(functionid ~ diet, scales="free",space="free") +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
              axis.text.y = element_text(size=6),
              strip.text.y = element_text(angle = 0)
              ) +
        labs(y="Traits",x="Samples",fill="GIFT")
```

```{r comunity_funct_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
unique_funct_db<- GIFT_db[c(3,4,5)] %>% 
  distinct(Code_function, .keep_all = TRUE)

GIFTs_functions_community %>%
   t() %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "Code_function")  %>%
  left_join(.,unique_funct_db[c(1,3)],by = join_by(Code_function == Code_function))  %>%
  select(-Code_function) %>%
  column_to_rownames(., "Function")%>%
   t()  %>%
    as.data.frame() %>%
    rownames_to_column(var="sample") %>%
    pivot_longer(!sample,names_to="trait",values_to="gift") %>%
    left_join(sample_metadata, by = join_by(sample == sample))%>%
 # filter(!diet =="Post_grass") %>%
    ggplot(aes(x=trait,y=sample,fill=gift)) +
        geom_tile(colour="white", linewidth=0.2)+
        scale_fill_gradientn(colours=rev(c("#d53e4f", "#f46d43", "#fdae61", "#fee08b", "#e6f598", "#abdda4", "#ddf1da")))+
        facet_grid(diet ~ ., scales="free",space="free")+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10),
        axis.text.y = element_text(size=8),
        strip.background = element_blank(),
        strip.text = element_text(size = 12, color="black",face="bold"),
        axis.title = element_text(size = 12, face = "bold"),
        axis.title.y = element_text(margin = margin(t = 0, r = 20, b = 0, l = 0)),
        axis.title.x = element_text(margin = margin(t = 20, r = 0, b = 0, l = 0)),
        panel.background= element_blank()
              ) +
        labs(x="Function", y="Sample",fill="GIFT")
```

## ANCOM-BC2

### Captive vs Wild 

#### Element level
```{r phyloseq_ele, comment="", echo=FALSE, message=FALSE, warning=FALSE}

sample_info_tab_phy <- sample_metadata%>%
  column_to_rownames(var="sample")%>%
  filter(diet!="Post_grass") %>% 
  sample_data()

count_phy <- GIFTs_elements_community %>% 
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  filter(sample %in% rownames(sample_info_tab_phy)) %>%
  column_to_rownames("sample") %>% 
  select(which(!colSums(., na.rm=TRUE) %in% 0)) %>% 
  t() %>%
  otu_table(., taxa_are_rows=T)

uniqueGIFT_db<- unique(GIFT_db[c(2,4,5,6)]) %>% unite("Function",Function:Element, sep= "_", remove=FALSE)

TAX <- uniqueGIFT_db%>%
    remove_rownames()%>%
  column_to_rownames(var="Code_element")%>%
  as.matrix()%>%
  tax_table()

physeq_function = phyloseq(count_phy, TAX, sample_info_tab_phy)
  
```

```{r ancom_rand_elem_capwild, comment="", echo=FALSE, message=FALSE, warning=FALSE, eval=FALSE}
set.seed(1234) #set seed for reproducibility
ancom_rand_output_element_captive_wild = ancombc2(data = physeq_function, 
                  assay_name = "counts", 
                  tax_level = NULL, #change to agglomerate analysis to a higher taxonomic range
                  fix_formula = "region", #fixed variable(s)
                  p_adj_method = "holm", 
                  pseudo_sens = TRUE,
                  prv_cut = 0, 
                  lib_cut = 0, 
                  s0_perc = 0.05,
                  group = NULL, 
                  struc_zero = FALSE, 
                  neg_lb = FALSE,
                  alpha = 0.05, 
                  n_cl = 2, 
                  verbose = TRUE,
                  global = FALSE, 
                  pairwise = FALSE, 
                  dunnet = FALSE, 
                  trend = FALSE,
                  iter_control = list(tol = 1e-5, max_iter = 20, verbose = FALSE),
                  em_control = list(tol = 1e-5, max_iter = 100),
                  mdfdr_control = list(fwer_ctrl_method = "holm", B = 100), 
                  trend_control = NULL)

```

```{r table_ele1, comment="", echo=FALSE, message=FALSE, warning=FALSE}
tax <- data.frame(physeq_function@tax_table) %>%
  rownames_to_column(., "taxon")
```

```{r ancom_rand_res_elem1, echo=FALSE, comment="", message=FALSE, warning=FALSE}
ancombc_rand_table <- ancom_rand_output_element_captive_wild$res %>%
  dplyr::select(taxon, lfc_regionNafarroa, p_regionNafarroa) %>%
  filter(p_regionNafarroa < 0.05) %>%
  dplyr::arrange(p_regionNafarroa) %>%
  merge(., tax, by="taxon")
```

```{r ancombc_rand_plot_elem1, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
ancombc_rand_table%>%
#      mutate(Function=factor(Function,levels=ancombc_rand_table$Function)) %>%
mutate(Color = ifelse(lfc_regionNafarroa <0, "Wild","Captive")) %>%
ggplot(aes(x=forcats::fct_reorder(Function,lfc_regionNafarroa), y=lfc_regionNafarroa, fill=Color)) + 
  geom_col() +
#  geom_point(size=4) + 
  scale_fill_manual(values=c("#e5bd5b", "#6b7398")) + 
  geom_hline(yintercept=0) + 
  coord_flip()+
  theme(axis.text = element_text(size = 10, face="bold.italic"),
        axis.title = element_text(size = 12),
        legend.position = "right", 
        legend.title = element_blank(),
        panel.background = element_blank(),
          panel.grid.major = element_line(size = 0.15, linetype = 'solid',
                                colour = "grey"))+
  xlab("Traits") + 
  ylab("log2FoldChange")
```

## Wilcoxon comparison

### Community elements differences:
```{r comunity_elem, comment="", echo=FALSE, message=FALSE, warning=FALSE}
element_gift <- GIFTs_elements_community %>% 
  as.data.frame() %>% 
  rownames_to_column(., "sample") %>% 
  left_join(sample_metadata[c(1,21)], by="sample") %>% 
  filter(diet!="Post_grass")
```

```{r commun_wilcox_elem, comment="", echo=FALSE, message=FALSE, warning=FALSE}
uniqueGIFT_db<- unique(GIFT_db[c(2,4,5,6)]) %>% unite("Function",Function:Element, sep= "_", remove=FALSE)

significant_elements <- element_gift %>%
    pivot_longer(-c(sample,diet), names_to = "trait", values_to = "value") %>%
    group_by(trait) %>%
    summarise(p_value = wilcox.test(value ~ diet)$p.value) %>%
    mutate(p_adjust=p.adjust(p_value, method="BH")) %>%
    filter(p_adjust < 0.05)%>%
  rownames_to_column(., "Elements")  %>% 
  select(-Elements)
# %>%
#   left_join(.,uniqueGIFT_db[c(1,3)],by = join_by(trait == Code_element))

element_gift_t <- element_gift  %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "trait")

element_gift_filt <- subset(element_gift_t, trait %in% significant_elements$trait) %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "sample")%>% 
  left_join(., sample_metadata[c(1,21)], by = join_by(sample == sample))

difference_table <- element_gift_filt %>%
  select(-sample) %>%
  group_by(diet) %>%
  summarise(across(everything(), mean)) %>%
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric) %>%
  rownames_to_column(., "Elements") %>%
  left_join(.,uniqueGIFT_db[c(1,3,4)],by = join_by(Elements == Code_element)) %>% 
  arrange(Function) %>% 
  mutate(Difference=Pre_grass-Wild)%>% 
  mutate(group_color = ifelse(Difference <0, "Wild","Captive")) 
```

```{r log_fold_calc, comment="", echo=FALSE, message=FALSE, warning=FALSE}
means_gift <- element_gift_filt %>% 
  select(-diet) %>% 
  pivot_longer(!sample, names_to = "elements", values_to = "abundance") %>% 
  left_join(sample_metadata, by=join_by(sample==sample)) %>% 
  group_by(diet, elements) %>%
  summarise(mean=mean(abundance))

log_fold <- means_gift %>%
  group_by(elements) %>%
  summarise(
    logfc_captive_wild = log2(mean[diet == "Pre_grass"] / mean[diet == "Wild"])
    )
```


```{r plot1, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE, eval=FALSE}
difference_table %>%
#  mutate(Difference_abs = abs(Difference)) %>% 
  left_join(significant_elements, by=join_by(Elements==trait)) %>% 
  ggplot(., aes(x = Difference, y = -log(p_adjust), color=group_color)) +
  geom_point(size=2, show.legend = FALSE) + 
  scale_color_manual(values=c("#e5bd5b", "#6b7398"))+
  #xlim(c(-10,4)) +
  theme_classic()+
  geom_text_repel(aes(label = Elements), size=2)
```

```{r plot3, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=5, fig.width=10, fig.fullwidth=TRUE, eval=FALSE}
uniqueGIFT <- unique(GIFT_db[c(2,3,4,5,6)])

code_function <- difference_table %>%
  left_join(uniqueGIFT[c(1:3)], by=join_by(Elements==Code_element))

unique_codes<-unique(code_function$Code_function)

gift_colors <- read_tsv("data/gift_colors.tsv") %>% 
  filter(Code_function %in% unique_codes)%>% 
  mutate(legend=str_c(Code_function," - ",Function))

code_function %>%
#  mutate(Difference_abs = abs(Difference)) %>% 
  left_join(significant_elements, by=join_by(Elements==trait)) %>%
  left_join(log_fold, by=join_by(Elements==elements)) %>% 
  left_join(gift_colors, by=join_by(Code_function==Code_function)) %>% 
  ggplot(., aes(x = logfc_captive_wild, y = -log(p_adjust), color=legend, size=abs(Difference))) +
  geom_point()+
  geom_vline(xintercept=0) +
  scale_color_manual(values = gift_colors$Color)+
  #xlim(c(-10,4)) +
  theme_classic()+
  labs(size="Mean difference (abs)", color="Functional trait")+
  labs(x = "Log-fold change", y="-Log adjusted p-value")

# +
#   geom_text_repel(aes(label = Elements), size=2, max.overlaps = Inf)

```
```{r final_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=8, fig.width=12, fig.fullwidth=TRUE}
code_function %>%
#  mutate(Difference_abs = abs(Difference)) %>% 
  left_join(significant_elements, by=join_by(Elements==trait)) %>%
  left_join(log_fold, by=join_by(Elements==elements)) %>% 
  left_join(gift_colors, by=join_by(Code_function==Code_function)) %>% 
  ggplot(., aes(x = logfc_captive_wild, y = -log(p_adjust), color=legend, size=abs(Difference))) +
  geom_jitter(width = 0.2, height = 0.2)+
  geom_vline(xintercept=0) +
  scale_color_manual(values = gift_colors$Color)+
  #xlim(c(-10,4)) +
  theme_classic()+
  labs(size="Mean difference (abs)", color="Functional trait")+
  labs(x = "Log-fold change", y="-Log adjusted p-value") +
  geom_text_repel(aes(label = Element), min.segment.length = 0.4, size=2.5, max.overlaps = Inf)
```

```{r plot4, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE, eval=FALSE}
difference_table %>%
#  mutate(Difference_abs = abs(Difference)) %>% 
  left_join(significant_elements, by=join_by(Elements==trait)) %>%
  left_join(log_fold, by=join_by(Elements==elements)) %>% 
  ggplot(., aes(x = Difference, y = logfc_captive_wild, color=group_color)) +
  geom_point(size=2, show.legend = FALSE) + 
  scale_color_manual(values=c("#e5bd5b", "#6b7398"))+
  #xlim(c(-10,4)) +
  theme_classic()+
  geom_text_repel(aes(label = Elements), size=2, max.overlaps = Inf)
```

```{r commun_wilcox_elem_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=6, fig.width=12, fig.fullwidth=TRUE, eval=FALSE}
difference_table %>%
  ggplot(aes(x=forcats::fct_reorder(Function,Difference), y=Difference, fill=group_color)) + 
  geom_col() +
#  geom_point(size=4) + 
  scale_fill_manual(values=c("#e5bd5b", "#6b7398")) + 
  geom_hline(yintercept=0) + 
  coord_flip()+
  theme(axis.text = element_text(size = 10),
        axis.title = element_text(size = 12),
        legend.position = "right", 
        legend.title = element_blank(),
        panel.background = element_blank(),
          panel.grid.major = element_line(size = 0.15, linetype = 'solid',
                                colour = "grey"))+
  xlab("Function") + 
  ylab("Mean difference")
```
```{r}
difference_table %>% 
  filter(group_color=="Wild")
```
```{r sugar_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=6, fig.width=12, fig.fullwidth=TRUE}
sugar_function <- GIFT_db %>% 
  filter(Code_function=="D03") %>% 
  select(Code_element) %>% 
  pull()

GIFTs_elements_community %>% 
  as.data.frame() %>% 
  select(any_of(sugar_function)) %>% 
  rownames_to_column(., "sample") %>% 
  left_join(sample_metadata[c(1,21)], by=join_by(sample==sample)) %>% 
  select(-sample) %>% 
  pivot_longer(!diet,names_to="trait",values_to="gift") %>% 
  left_join(uniqueGIFT_db, by=join_by(trait==Code_element)) %>% 
  ggplot(aes(x=diet, y=gift, color=diet)) +
  geom_jitter(width = 0.2, show.legend = FALSE) +
  geom_boxplot(width = 0.5, alpha=0.5,outlier.shape = NA, show.legend = FALSE) +
  scale_color_manual(values=diet_colors)+
  scale_fill_manual(values=diet_colors) +
  scale_x_discrete(labels=c("Wild" = "Wild", "Pre_grass" = "Captive-born")) +
facet_grid(.~Element,  scales="free_x")+
  coord_cartesian(xlim = c(1, NA)) +
  theme_classic() +
  theme(
    strip.background = element_blank(),
    panel.grid.minor.x = element_line(size = .1, color = "grey"),
    axis.text = element_text(size=10, angle = 45, hjust = 1),
    axis.title.y = element_text(margin = margin(t = 0, r = 20, b = 0, l = 0)),
    axis.title.x = element_text(margin = margin(t = 20, r = 0, b = 0, l = 0))
    )
```

```{r sugar_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=6, fig.width=12, fig.fullwidth=TRUE}
poly_function <- GIFT_db %>% 
  filter(Code_function=="D02") %>% 
  select(Code_element) %>% 
  pull()

GIFTs_elements_community %>% 
  as.data.frame() %>% 
  select(any_of(poly_function)) %>% 
  rownames_to_column(., "sample") %>% 
  left_join(sample_metadata[c(1,21)], by=join_by(sample==sample)) %>% 
  select(-sample) %>% 
  pivot_longer(!diet,names_to="trait",values_to="gift") %>% 
  left_join(uniqueGIFT_db, by=join_by(trait==Code_element)) %>% 
  ggplot(aes(x=diet, y=gift, color=diet)) +
  geom_jitter(width = 0.2, show.legend = FALSE) +
  geom_boxplot(width = 0.5, alpha=0.5,outlier.shape = NA, show.legend = FALSE) +
  scale_color_manual(values=diet_colors)+
  scale_fill_manual(values=diet_colors) +
  scale_x_discrete(labels=c("Wild" = "Wild", "Pre_grass" = "Captive-born")) +
facet_grid(.~Element,  scales="free_x")+
  coord_cartesian(xlim = c(1, NA)) +
  theme_classic() +
  theme(
    strip.background = element_blank(),
    panel.grid.minor.x = element_line(size = .1, color = "grey"),
    axis.text = element_text(size=10, angle = 45, hjust = 1),
    axis.title.y = element_text(margin = margin(t = 0, r = 20, b = 0, l = 0)),
    axis.title.x = element_text(margin = margin(t = 20, r = 0, b = 0, l = 0))
    )
```

```{r elements_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE}
element_gift_names <- element_gift_filt%>%
  select(-diet)%>%
   t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "Elements")  %>%
  left_join(.,uniqueGIFT_db[c(1,3)],by = join_by(Elements == Code_element))%>%
  select(-Elements)%>%
  select(Function, everything())%>%
   t()%>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "sample")%>% 
  left_join(., sample_metadata[c(1,21)], by = join_by(sample == sample))


colNames <- names(element_gift_names)[2:30]
for(i in colNames){
  plt <- ggplot(element_gift_names, aes(x=diet, y=.data[[i]], color = diet, fill=diet)) +
    geom_boxplot(alpha = 0.2, outlier.shape = NA, width = 0.3, show.legend = FALSE) +
  geom_jitter(width = 0.1, show.legend = TRUE) +
    scale_color_manual(values=c("#e5bd5b", "#6b7398"))+
    scale_fill_manual(values=c("#e5bd5b", "#6b7398"))+
  theme_minimal() +
  theme(
    axis.line = element_line(size = 0.5, linetype = "solid", colour = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())
print(plt)
}

```








#### Functional level
```{r phylo_func1, comment="", echo=FALSE, message=FALSE, warning=FALSE}
sample_info_tab_phy <- sample_metadata%>%
  column_to_rownames(var="sample")%>%
  filter(diet!="Post_grass") %>%
  sample_data()

count_phy <- GIFTs_functions_community %>% 
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  filter(sample %in% rownames(sample_info_tab_phy)) %>%
  column_to_rownames("sample") %>% 
  select(which(!colSums(., na.rm=TRUE) %in% 0)) %>% 
  t() %>%
  otu_table(., taxa_are_rows=T)

TAX <- unique_funct_db%>%
    remove_rownames()%>%
  column_to_rownames(var="Code_function")%>%
  as.matrix()%>%
  tax_table()

physeq_functional_filtered = phyloseq(count_phy, TAX, sample_info_tab_phy)
physeq_functional_filtered_clr <- microbiome::transform(physeq_functional_filtered, 'clr')  
```


```{r ancom_rand_func_prewild, comment="", echo=FALSE, message=FALSE, warning=FALSE, eval=FALSE}
set.seed(1234) #set seed for reproducibility
ancom_rand_output_function_pre_wild = ancombc2(data = physeq_functional_filtered, 
                  assay_name = "counts", 
                  tax_level = NULL, #change to agglomerate analysis to a higher taxonomic range
                  fix_formula = "region", #fixed variable(s)
                  p_adj_method = "holm", 
                  pseudo_sens = TRUE,
                  prv_cut = 0, 
                  lib_cut = 0, 
                  s0_perc = 0.05,
                  group = NULL, 
                  struc_zero = FALSE, 
                  neg_lb = FALSE,
                  alpha = 0.05, 
                  n_cl = 2, 
                  verbose = TRUE,
                  global = FALSE, 
                  pairwise = FALSE, 
                  dunnet = FALSE, 
                  trend = FALSE,
                  iter_control = list(tol = 1e-5, max_iter = 20, verbose = FALSE),
                  em_control = list(tol = 1e-5, max_iter = 100),
                  mdfdr_control = list(fwer_ctrl_method = "holm", B = 100), 
                  trend_control = NULL)

```

```{r table_func1, comment="", echo=FALSE, message=FALSE, warning=FALSE}
tax <- data.frame(physeq_functional_filtered@tax_table) %>%
  rownames_to_column(., "taxon")

ancombc_rand_func <- ancom_rand_output_function_pre_wild$res %>%
  dplyr::select(taxon, lfc_regionNafarroa, p_regionNafarroa) %>%
  filter(p_regionNafarroa < 0.05) %>%
  dplyr::arrange(p_regionNafarroa) %>%
  merge(., tax, by="taxon") %>%
  dplyr::arrange(lfc_regionNafarroa)
```

```{r ancom_rand_res_funct1, comment="", echo=FALSE, message=FALSE, warning=FALSE}
ancombc_rand_table_func <- ancom_rand_output_function_pre_wild$res %>%
  dplyr::select(taxon, lfc_regionNafarroa, p_regionNafarroa) %>%
  filter(p_regionNafarroa < 0.05) %>%
  dplyr::arrange(p_regionNafarroa) %>%
  merge(., tax, by="taxon")
```

```{r ancombc_rand_plot_funct, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=4, fig.width=8, fig.fullwidth=TRUE}
ancombc_rand_table_func%>%
mutate(Color = ifelse(lfc_regionNafarroa <0, "Wild","Captive")) %>%
ggplot(aes(x=forcats::fct_reorder(Function,lfc_regionNafarroa), y=lfc_regionNafarroa, fill=Color)) + 
  geom_col() +
  scale_fill_manual(values=c("#e5bd5b", "#6b7398")) + 
  geom_hline(yintercept=0) + 
  coord_flip()+
  theme(axis.text = element_text(size = 10, face="bold.italic"),
        axis.title = element_text(size = 12),
        legend.position = "right", 
        legend.title = element_blank(),
        panel.background = element_blank(),
          panel.grid.major = element_line(size = 0.15, linetype = 'solid',
                                colour = "grey"))+
  xlab("Traits") + 
  ylab("log2FoldChange")
```


### Before and after grass

#### Element level
```{r phyloseq_ele_post, comment="", echo=FALSE, message=FALSE, warning=FALSE}

sample_info_tab_phy <- sample_metadata%>%
  column_to_rownames(var="sample")%>%
  filter(diet!="Wild") %>% 
  sample_data()

count_phy <- GIFTs_elements_community %>% 
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  filter(sample %in% rownames(sample_info_tab_phy)) %>%
  column_to_rownames("sample") %>% 
  select(which(!colSums(., na.rm=TRUE) %in% 0)) %>% 
  t() %>%
  otu_table(., taxa_are_rows=T)

uniqueGIFT_db<- unique(GIFT_db[c(2,4,5,6)]) %>% unite("Function",Function:Element, sep= "_", remove=FALSE)

TAX <- uniqueGIFT_db%>%
    remove_rownames()%>%
  column_to_rownames(var="Code_element")%>%
  as.matrix()%>%
  tax_table()

physeq_function = phyloseq(count_phy, TAX, sample_info_tab_phy)
  
```

```{r structural_functional_capacity, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=12, fig.width=10, fig.fullwidth=TRUE}

ancombc_rand_table_mag <- ancombc_rand_table_mag %>% 
  mutate(group = if_else(lfc_dietPre_grass > 0, "Pre_grass", "Post_grass"))
GIFTs_elements_filtered %>% 
  rownames_to_column("genome") %>%
  filter(genome %in% ancombc_rand_table_mag$taxon)  %>% 
    pivot_longer(!genome,names_to="trait",values_to="gift") %>% 
    left_join(ancombc_rand_table_mag, by = join_by(genome == taxon))%>%
    mutate(functionid = substr(trait, 1, 3)) %>%
    mutate(trait = case_when(
      trait %in% GIFT_db$Code_element ~ GIFT_db$Element[match(trait, GIFT_db$Code_element)],
      TRUE ~ trait
    )) %>%
    mutate(functionid = case_when(
      functionid %in% GIFT_db$Code_function ~ GIFT_db$Function[match(functionid, GIFT_db$Code_function)],
      TRUE ~ functionid
    )) %>%
    mutate(trait=factor(trait,levels=unique(GIFT_db$Element))) %>%
    mutate(functionid=factor(functionid,levels=unique(GIFT_db$Function))) %>%
    ggplot(aes(x=genome,y=trait,fill=gift)) +
        geom_tile(colour="white", linewidth=0.2)+
        scale_fill_gradientn(colours=rev(c("#d53e4f", "#f46d43", "#fdae61", "#fee08b", "#e6f598", "#abdda4", "#ddf1da")))+
        facet_grid(functionid ~ group, scales="free",space="free") +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
              axis.text.y = element_text(size=6),
              strip.text.y = element_text(angle = 0)
              ) +
        labs(y="Traits",x="Samples",fill="GIFT")

```

```{r ancom_rand_elem_post, comment="", echo=FALSE, message=FALSE, warning=FALSE, eval=FALSE}
set.seed(1234) #set seed for reproducibility
ancom_rand_output_element_pre_post = ancombc2(data = physeq_function, 
                  assay_name = "counts", 
                  tax_level = NULL, #change to agglomerate analysis to a higher taxonomic range
                  fix_formula = "diet", #fixed variable(s)
                  rand_formula = "(1|individual)",
                  p_adj_method = "holm", 
                  pseudo_sens = TRUE,
                  prv_cut = 0, 
                  lib_cut = 0, 
                  s0_perc = 0.05,
                  group = NULL, 
                  struc_zero = FALSE, 
                  neg_lb = FALSE,
                  alpha = 0.05, 
                  n_cl = 2, 
                  verbose = TRUE,
                  global = FALSE, 
                  pairwise = FALSE, 
                  dunnet = FALSE, 
                  trend = FALSE,
                  iter_control = list(tol = 1e-5, max_iter = 20, verbose = FALSE),
                  em_control = list(tol = 1e-5, max_iter = 100),
                  lme_control = lme4::lmerControl(),
                  mdfdr_control = list(fwer_ctrl_method = "holm", B = 100), 
                  trend_control = NULL)

```

```{r table_ele2, comment="", echo=FALSE, message=FALSE, warning=FALSE}
tax <- data.frame(physeq_function@tax_table) %>%
  rownames_to_column(., "taxon")

ancombc_rand_elem <- ancom_rand_output_element_pre_post$res %>%
  dplyr::select(taxon, lfc_dietPre_grass, p_dietPre_grass) %>%
  filter(p_dietPre_grass < 0.05) %>%
  dplyr::arrange(p_dietPre_grass) %>%
  merge(., tax, by="taxon") %>%
  dplyr::arrange(lfc_dietPre_grass)

```

```{r ancom_rand_res_elem2, echo=FALSE, comment="", message=FALSE, warning=FALSE}
ancombc_rand_table <- ancom_rand_output_element_pre_post$res %>%
  dplyr::select(taxon, lfc_dietPre_grass, p_dietPre_grass) %>%
  filter(p_dietPre_grass < 0.05) %>%
  dplyr::arrange(p_dietPre_grass) %>%
  merge(., tax, by="taxon")
```

```{r ancombc_rand_plot_elem2, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
ancombc_rand_table%>%
#      mutate(Function=factor(Function,levels=ancombc_rand_table$Function)) %>%
mutate(Color = ifelse(lfc_dietPre_grass <0, "Post","Pre")) %>%
ggplot(aes(x=forcats::fct_reorder(Function,lfc_dietPre_grass), y=lfc_dietPre_grass, fill=Color)) + 
  geom_col() +
#  geom_point(size=4) + 
  scale_fill_manual(values=c("#e5bd5b", "#6b7398")) + 
  geom_hline(yintercept=0) + 
  coord_flip()+
  theme(axis.text = element_text(size = 10, face="bold.italic"),
        axis.title = element_text(size = 12),
        legend.position = "right", 
        legend.title = element_blank(),
        panel.background = element_blank(),
          panel.grid.major = element_line(size = 0.15, linetype = 'solid',
                                colour = "grey"))+
  xlab("Traits") + 
  ylab("log2FoldChange")
```

#### Functional level
```{r phylo_func2, comment="", echo=FALSE, message=FALSE, warning=FALSE}

sample_info_tab_phy <- sample_metadata%>%
  column_to_rownames(var="sample")%>%
  filter(diet!="Wild") %>%
  sample_data()

count_phy <- GIFTs_functions_community %>% 
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  filter(sample %in% rownames(sample_info_tab_phy)) %>%
  column_to_rownames("sample") %>% 
  select(which(!colSums(., na.rm=TRUE) %in% 0)) %>% 
  t() %>%
  otu_table(., taxa_are_rows=T)

TAX <- unique_funct_db%>%
    remove_rownames()%>%
  column_to_rownames(var="Code_function")%>%
  as.matrix()%>%
  tax_table()

physeq_functional_filtered = phyloseq(count_phy, TAX, sample_info_tab_phy)
physeq_functional_filtered_clr <- microbiome::transform(physeq_functional_filtered, 'clr')  
```


```{r ancom_rand_func2, comment="", echo=FALSE, message=FALSE, warning=FALSE, eval=FALSE}
set.seed(1234) #set seed for reproducibility
ancom_rand_output_function_pre_post = ancombc2(data = physeq_functional_filtered, 
                  assay_name = "counts", 
                  tax_level = NULL, #change to agglomerate analysis to a higher taxonomic range
                  fix_formula = "diet", #fixed variable(s)
                  rand_formula = "(1|individual)",
                  p_adj_method = "holm", 
                  pseudo_sens = TRUE,
                  prv_cut = 0, 
                  lib_cut = 0, 
                  s0_perc = 0.05,
                  group = NULL, 
                  struc_zero = FALSE, 
                  neg_lb = FALSE,
                  alpha = 0.05, 
                  n_cl = 2, 
                  verbose = TRUE,
                  global = FALSE, 
                  pairwise = FALSE, 
                  dunnet = FALSE, 
                  trend = FALSE,
                  iter_control = list(tol = 1e-5, max_iter = 20, verbose = FALSE),
                  em_control = list(tol = 1e-5, max_iter = 100),
                  lme_control = lme4::lmerControl(),
                  mdfdr_control = list(fwer_ctrl_method = "holm", B = 100), 
                  trend_control = NULL)

```
```{r table_func2, comment="", echo=FALSE, message=FALSE, warning=FALSE}
tax <- data.frame(physeq_functional_filtered@tax_table) %>%
  rownames_to_column(., "taxon")

ancombc_rand_func <- ancom_rand_output_function_pre_post$res %>%
  dplyr::select(taxon, lfc_dietPre_grass, p_dietPre_grass) %>%
  filter(p_dietPre_grass < 0.05) %>%
  dplyr::arrange(p_dietPre_grass) %>%
  merge(., tax, by="taxon") %>%
  dplyr::arrange(lfc_dietPre_grass)
ancombc_rand_func
```


```{r ancom_save_results, comment="", echo=FALSE, message=FALSE, warning=FALSE, eval=FALSE}
save(ancom_rand_output_element_captive_wild,
     ancom_rand_output_function_pre_wild,
     ancom_rand_output_element_pre_post,
     ancom_rand_output_function_pre_post,
     file="data/ancombc_filtered_functional.Rdata")
```


## CompleteR

```{r}
library(devtools)
library(completer)
library(phyloseq)
library(tidyverse)
#setup_completer()
```


```{r}
lepus_tree <- read_tree("data/DMB0165_gtdbtk.bac120.classify.tree")
lepus_tree$tip.label <- str_replace_all(lepus_tree$tip.label,"'", "") #remove single quotes in MAG names
lepus_tree <- keep.tip(lepus_tree, tip=genome_metadata$genome)

lepus_annotations <- read_tsv("data/genome_annotations.tsv.xz")
lepus_traits <- annot_to_traits(annot=lepus_annotations,genome_index=2,kegg_index=9)
lepus_imputed <- completer(traits = lepus_traits, tree = lepus_tree)
```

